cdc/owner/ddl_puller.go

```diff
@@ -0,0 +1,168 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package owner
+
+import (
+	"context"
+	"sync"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	timodel "github.com/pingcap/parser/model"
+	"github.com/pingcap/ticdc/cdc/entry"
+	"github.com/pingcap/ticdc/cdc/model"
+	"github.com/pingcap/ticdc/cdc/puller"
+	"github.com/pingcap/ticdc/pkg/config"
+	cdcContext "github.com/pingcap/ticdc/pkg/context"
+	"github.com/pingcap/ticdc/pkg/filter"
+	"github.com/pingcap/ticdc/pkg/regionspan"
+	"github.com/pingcap/ticdc/pkg/util"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+// DDLPuller is a wrapper of the Puller interface for the owner
+// DDLPuller start a puller and listen to the DDL range, add the received DDL into an internal queue
+type DDLPuller interface {
+	// Run runs the DDLPuller
+	Run(ctx cdcContext.Context) error
+	// FrontDDL returns the first DDL job in the internal queue
+	FrontDDL() (uint64, *timodel.Job)
+	// PopFrontDDL returns and pops the first DDL job in the internal queue
+	PopFrontDDL() (uint64, *timodel.Job)
+	// Close closes the DDLPuller
+	Close()
+}
+
+type ddlPullerImpl struct {
+	puller puller.Puller
+	filter *filter.Filter
+
+	mu             sync.Mutex
+	resolvedTS     uint64
+	pendingDDLJobs []*timodel.Job
+	cancel         context.CancelFunc
+}
+
+func newDDLPuller(ctx cdcContext.Context, startTs uint64) (DDLPuller, error) {
+	pdCli := ctx.GlobalVars().PDClient
+	conf := config.GetGlobalServerConfig()
+	f, err := filter.NewFilter(ctx.ChangefeedVars().Info.Config)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	var plr puller.Puller
+	kvStorage := ctx.GlobalVars().KVStorage
+	// kvStorage can be nil only in the test
+	if kvStorage != nil {
+		plr = puller.NewPuller(ctx, pdCli, conf.Security, kvStorage, startTs,
+			[]regionspan.Span{regionspan.GetDDLSpan(), regionspan.GetAddIndexDDLSpan()},
+			nil, false)
+	}
+
+	return &ddlPullerImpl{
+		puller: plr,
+		// the puller will listen changed events from `startTs`(including `startTs`)
+		// so `startTs - 1` is a available resolvedTS, it means that all txn before `startTs - 1` is received(or don't need to receive)
+		resolvedTS: startTs - 1,
+		filter:     f,
+		cancel:     func() {},
+	}, nil
+}
+
+const ddlPullerName = "DDL_PULLER"
+
+func (h *ddlPullerImpl) Run(ctx cdcContext.Context) error {
+	ctx, cancel := cdcContext.WithCancel(ctx)
+	h.cancel = cancel
+	log.Debug("DDL puller started")
+	stdCtx := util.PutTableInfoInCtx(ctx, -1, ddlPullerName)
+	errg, stdCtx := errgroup.WithContext(stdCtx)
+	ctx = cdcContext.WithStd(ctx, stdCtx)
+
+	errg.Go(func() error {
+		return h.puller.Run(ctx)
+	})
+
+	rawDDLCh := puller.SortOutput(ctx, h.puller.Output())
+
+	receiveDDL := func(rawDDL *model.RawKVEntry) error {
+		if rawDDL == nil {
+			return nil
```

Comment:
Is this branch ever reachable in production?

Reply:
should never reachable, just for safe